deepcell_toolbox/utils.py

```diff
@@ -309,12 +311,37 @@ def resize(data, shape, data_format='channels_last'):
         else:
             shape = tuple(list(shape) + [data.shape[channel_axis]])
 
-        _resize = lambda d: transform.resize(d, shape, mode='constant', preserve_range=True)
+        # linear interpolation (order 1) for image data
+        if data_type == 'X':
+            _resize = lambda d: transform.resize(d, shape, mode='constant', preserve_range=True,
+                                                 order=1)
+
+        # nearest neighbors (order 0) for labels data, but have to use cv2 to avoid floats
+        else:
+            def _resize(d):
+                resized_labels = []
+                for i in range(data.shape[channel_axis]):
+                    if data_format == 'channels_first':
+                        resized_label = cv2.resize(d[i, ...], shape[1:],
+                                                   interpolation=cv2.INTER_NEAREST)
+                    else:
+                        resized_label = cv2.resize(d[..., i], shape[:-1],
+                                                   interpolation=cv2.INTER_NEAREST)
```

Comment:
why does `data_format` only matter for integer data?

Reply:
skimage nearest neighbor interpolation produces floats when downsampling, so I had to use cv2 instead.

Reply:
What happens if you just cast back to an int after it becomes a float? I would think int -> float -> would have no issue.

Reply:
I would try casting back to int before relying on cv2 in that particular case.

Reply:
Yeah, that's what I tried below. But then it creates integer interpolated values between labels.

Reply:
Can you try these settings? [This](https://github.com/vanvalenlab/kiosk-redis-consumer/blob/5df00440d69ddb3aea2562b708a6f44bf68220b8/redis_consumer/utils.py#L455) is what we were using in the redis_consumer 
```
rescaled_img = skimage.transform.rescale(
        image, scale,
        mode='edge',
        anti_aliasing=False,
        anti_aliasing_sigma=None,
        multichannel=multichannel,
        preserve_range=True,
        order=0
    )
```

Reply:
Oooh awesome, I didn't know `rescale` existed! That definitely does what we need it to do for the labels. 

However, it doesn't take an output size, but rather a scale factor. If I modify [the test function](https://github.com/vanvalenlab/kiosk-redis-consumer/blob/5df00440d69ddb3aea2562b708a6f44bf68220b8/redis_consumer/utils_test.py#L379) to instead take in random scales: 

```
def test_rescale():
    scales = np.random.rand(20)
    shapes = [(4, 4, 5), (4, 4, 1), (4, 4)]
    for scale in scales:
        for shape in shapes:
            image = np.random.random(shape)
            rescaled = utils.rescale(image, 1)
            np.testing.assert_array_equal(rescaled, image)

            rescaled = utils.rescale(image, scale)
            expected_shape = (int(np.ceil(shape[0] * scale)),
                              int(np.ceil(shape[1] * scale)))

            if len(shape) > 2:
                expected_shape = tuple(list(expected_shape) + [int(shape[2])])
            
            # commented out to see if final output shape is same as input
            # assert rescaled.shape == expected_shape
            # scale it back
            rescaled = utils.rescale(rescaled, 1 / scale)
            assert rescaled.shape == shape
```

It fails due to off-by-one pixel errors. Do we want to modify the logic [in the applications resize function](https://github.com/vanvalenlab/deepcell-tf/blob/890b7cd85983eb2811c5b0689431267df6e5f66e/deepcell/applications/application.py#L106) to only pass friendly resize ratios?

I was thinking that modifying it to have a range in which it's close enough and doesn't resize might be a good idea anyway?

Reply:
Okay the issue was the anti-aliasing. If I use resize (which lets us specify precise input/output pixel sizes) with the parameters above (anti-aliasing=False) then we get the behavior we want!

I'll update the PR